### PR TITLE
gh-120323: Constant fold entire attribute loads

### PR DESCRIPTION
**Pull request by Fidget-Spinner**
From the PR's test case. This constant folds entire code like:
```python
class MyEnum:
    A = 1

def testfunc(n):
    for i in range(n):
        x = MyEnum.A
        y = MyEnum.A
```

To 
```python
class MyEnum:
    A = 1

def testfunc(n):
    for i in range(n):
        x = MyEnum.A
        y = 1
```
Using watchers.

At the moment, it's really limited and only constant folds simple class loads. In the future we can constant fold method descriptors and more.

* Issue: gh-120323

### CODE REVIEW

File: Python/optimizer_symbols.c [hunk 1]
```diff
@@ -188,6 +196,20 @@ _Py_uop_sym_set_const(_Py_UOpsContext *ctx, _Py_UopsSymbol *sym, PyObject *const
         sym->typ = typ;
         sym->const_val = Py_NewRef(const_val);
     }
+
+    if (PyType_Check(const_val)) {
+        sym_set_flag(sym, IS_TYPE_SUBCLASS);
+        _Py_BloomFilter_Add(ctx->dependencies, const_val);
+        PyType_Watch(TYPE_WATCHER_ID, (PyObject *)const_val);
```
Comment (markshannon) [inline Python/optimizer_symbols.c]: Hidden side effects like this make debugging and maintenance *much* harder, as any code that calls `sym_set_const` can no longer be understood simply by reading it. 
`_Py_uop_sym_set_const` should set the symbol to be constant and nothing else.

For future reference, bear the following rules in mind:
* Does the name reflect *exactly* what the function does?
* If it does, and the function name contains "and" or "or", consider break it up.

So, in this case a "correct" name would be `sym_set_const_and_make_trace_depend_on_the_version_if_const_is_a_class`.
Which would, at least, warn users not to use it if the const were a class and they didn't want the dependency.

File: Python/optimizer_symbols.c [hunk 2]
```diff
@@ -122,6 +123,13 @@ _Py_uop_sym_is_null(_Py_UopsSymbol *sym)
     return sym->flags == IS_NULL;
 }
 
+bool
+_Py_uop_sym_is_type_subclass(_Py_UopsSymbol *sym)
```
Comment (markshannon) [inline Python/optimizer_symbols.c]:
```suggestion
_Py_uop_sym_is_a_class(_Py_UopsSymbol *sym)
```

File: Python/bytecodes.c [hunk 3]
```diff
@@ -2052,7 +2055,8 @@ dummy_func(
 
         macro(LOAD_ATTR_CLASS) =
             unused/1 +
-            _CHECK_ATTR_CLASS +
```
Comment (markshannon) [inline Python/bytecodes.c]: Is it worth breaking this up?
It will slow down code that needs it, and it looks like the optimizer will remove both of the parts in most cases.

File: Python/optimizer_bytecodes.c [hunk 4]
```diff
@@ -117,6 +117,10 @@ dummy_func(void) {
 
     op(_GUARD_TYPE_VERSION, (type_version/2, owner -- owner)) {
         assert(type_version);
+        if (sym_is_type_subclass(owner)) {
```
Comment (markshannon) [inline Python/optimizer_bytecodes.c]: I don't think this should be necessary.
If this case isn't handled correctly by `sym_matches_type_version` and `sym_set_type_version` below, then those need fixing.

File: Python/optimizer_analysis.c [hunk 5]
```diff
@@ -489,6 +486,65 @@ optimize_uops(
 
 }
 
+static inline bool
+op_is_simple_load(int opcode) {
```
Comment (markshannon) [inline Python/optimizer_analysis.c:490]: Can we avoid mixing refactorings and adding new features?

`remove_unneeded_uops` coould do with cleaning up, but it should be done in another PR.

File: Python/optimizer_analysis.c [hunk 6]
```diff
@@ -489,6 +486,65 @@ optimize_uops(
...
+        case _LOAD_CONST_INLINE:
+        case _LOAD_FAST:
+            return true;
+        default:
+            return false;
+    }
+}
+
+static bool
+remove_simple_pops(int num_popped, _PyUOpInstruction *curr, _PyUOpInstruction *limit){
```
Comment (markshannon) [inline Python/optimizer_analysis.c:502]: If this removes simple pops, as its name suggests, why does it return a `bool`?

**Review by markshannon — CHANGES_REQUESTED** [review]
The optimization looks useful, and fits in nicely with other constant propagation/removal.

A couple of general rules to follow:

Avoid side effects in functions unless that function performs only that side effect.
Keep new features and refactoring in separate PRs, where possible.

File: Python/optimizer_symbols.c (hunk 1 above)
Comment (Fidget-Spinner) [reply Python/optimizer_symbols.c]: I removed it.

File: Python/optimizer_bytecodes.c (hunk 4 above)
Comment (Fidget-Spinner) [reply Python/optimizer_bytecodes.c]: We need this because it might have a type version that matches, but is a type subclass, not a normal object.

Comment (markshannon) [reply Python/optimizer_bytecodes.c]: `_GUARD_TYPE_VERSION` guards the version number of `owner`'s class. Whether `owner` is a class or not seems irrelevant.

File: Python/optimizer_bytecodes.c [hunk 7]
```diff
@@ -138,6 +142,24 @@ dummy_func(void) {
...
+    op(_CHECK_ATTR_CLASS, (type_version/2, owner -- owner)) {
+        assert(type_version);
+        if (sym_is_a_class(owner) &&
+            sym_matches_type_version(owner, type_version)) {
+            REPLACE_OP(this_instr, _NOP, 0, 0);
+        } else {
+            sym_set_is_a_class(owner);
+            PyTypeObject *type = _PyType_LookupByVersion(type_version);
+            if (type) {
+                if (sym_set_type_version(owner, type_version)) {
```
Comment (markshannon) [inline Python/optimizer_bytecodes.c:154]: This records that the object's class has the version number, not that the object  itself is a class and has a version number.

**Comment by markshannon** [comment]
AFAICT, this PR does a number of things which would be better in their own PRs.

1. Optimize away `_CHECK_ATTR_CLASS` using a type watcher.
2. Rename `_LOAD_ATTR_CLASS` to `_POP_TOP_LOAD_CONST_INLINE_WITH_NULL` and optimize to the "borrow" form.
3. Some ad-hoc improvements to the `remove_unneeded_uops`

Can you make these separate PRs, please?
2. could also include `_LOAD_ATTR_METHOD_WITH_VALUES` and any other uops that replace TOS with a constant (possibly pushing NULL).
